scipy/linalg/tests/test_lapack.py

```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
         else:
             assert_array_almost_equal(desired_cplx.astype(dtype),
                                       x/scale, decimal=4)
+
+
+def test_gttrf_gttrs():
+    #%%
+    np.random.seed(42)
+    # use ?gttrf and ?gttrs to solve a random linear system for each dtype
+    for index, dtype in enumerate(DTYPES):
+        n = 10
+        # set appropriate test tolerance based on the data type
+        if index % 2 == 0:
+            rtol = 10**-(np.finfo(dtype).precision-1)
+        else:
+            rtol = 10**-(np.finfo(dtype).precision-3)
+        # create the matrix in accordance with the data type
+        if index < 2:
+            du = np.random.rand(n-1)
+            d = np.random.rand(n)
+            dl = np.random.rand(n-1)
+        else:
+            du = np.random.rand(n-1) + np.random.rand(n-1) * 1j
+            d = np.random.rand(n) + np.random.rand(n) * 1j
+            dl = np.random.rand(n-1) + np.random.rand(n-1) * 1j
+
+        diag_cpy = np.array([dl.copy(), d.copy(), du.copy()])
+
+        A = np.diag(d) + np.diag(dl, -1) + np.diag(du, 1)
+        x = np.random.rand(n)
+        b = A @ x
+
+        gttrf, gttrs = get_lapack_funcs(('gttrf', 'gttrs'), dtype=dtype)
+
+        _dl, _d, _du, du2, ipiv, info = gttrf(dl, d, du)
+        # test to assure that the inputs of ?gttrf are maintained
+        assert_array_equal(dl, diag_cpy[0])
+        assert_array_equal(d, diag_cpy[1])
+        assert_array_equal(du, diag_cpy[2])
+
+        perm_fnl = np.arange(n)
+        for i, piv in enumerate(ipiv):
+            perm_fnl[i], perm_fnl[piv-1] = perm_fnl[piv-1], perm_fnl[i]
+
+        U = np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)
+        L = np.eye(n)
+
+        for i, m in enumerate(_dl):
+            piv = ipiv[i] - 1
+
+            # create permutation matrix
+            P = np.eye(n)
+            temp = P[i,:].copy()
+            P[i,:] = P[piv, :]
+            P[piv, :] = temp
+
+            # create factor Li
+            if index < 2:
+                Li = np.eye(n)
+            else:
+                Li = np.eye(n) + np.eye(n)*1j
+            Li[i+1,i] = m
+
+            L = L@P@Li
+
+        # one last permutation
+        piv = ipiv[-1] - 1
+        i = -1
+        P = np.eye(n)
+        temp = P[i,:].copy()
+        P[i,:] = P[piv, :]
+        P[piv,:] = temp
+
```

Comment:
```suggestion
        L = L@P
```
I've actually never seen the last element of `ipiv` be something other than `n`, but I think in general the last permutation is needed.

Reply:
Actually, you can replace the creation of the last permutation matrix and `L=L@P` with:
```
i, piv = -1, ipiv[-1] - 1
L[:,[i,piv]] = L[:,[piv,i]]  # right multiply by final permutation matrix
```